# Dependencies

In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np

# EDA

## Summary Dataset

In [ ]:
df = pd.read_csv('hotel_bookings_data.csv')
df.sample(5)

hotel  is_canceled  lead_time  arrival_date_year  \
99231     City Hotel            0        186               2018   
104622    City Hotel            0         14               2019   
33324   Resort Hotel            0         16               2019   
115200    City Hotel            0        126               2019   
15948   Resort Hotel            0         78               2017   

       arrival_date_month  arrival_date_week_number  \
99231            December                        41   
104622              March                         3   
33324               April                         7   
115200          September                        27   
15948             October                        32   

        arrival_date_day_of_month  stays_in_weekend_nights  \
99231                           6                        2   
104622                         16                        1   
33324                          17                        1   
115200                          2                        1   
15948                           2                        2   

        stays_in_weekdays_nights  adults  ...  booking_changes  deposit_type  \
99231                          3       2  ...                0    No Deposit   
104622                         2       2  ...                0    No Deposit   
33324                          2       1  ...                0    No Deposit   
115200                         0       2  ...                0    No Deposit   
15948                          5       2  ...                0    No Deposit   

        agent company days_in_waiting_list customer_type     adr  \
99231     7.0     NaN                    0      Personal   79.66   
104622    9.0     NaN                    0      Personal  103.67   
33324     NaN   498.0                    0        Family   36.00   
115200    8.0     NaN                    0      Personal  133.60   
15948   171.0     NaN                    0      Personal  101.65   

        required_car_parking_spaces  total_of_special_requests  \
99231                             0                          1   
104622                            0                          1   
33324                             0                          0   
115200                            0                          0   
15948                             0                          0   

        reservation_status  
99231            Check-Out  
104622           Check-Out  
33324            Check-Out  
115200           Check-Out  
15948            Check-Out  

[5 rows x 29 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 29 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_weekdays_nights        119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [ ]:
categorical = [
    'hotel', 'is_canceled', 'meal', 'city', 'market_segment',
    'distribution_channel', 'is_repeated_guest', 'reservation_status',
    'deposit_type', 'agent', 'company', 'customer_type'
]
numerical = [
    'adults', 'children', 'babies', 'previous_cancellations',
    'previous_bookings_not_canceled', 'booking_changes', 'adr',
    'required_car_parking_spaces', 'total_of_special_requests'
    
]
time_related = [
    'lead_time', 'arrival_date_year', 'arrival_date_month',
    'arrival_date_week_number', 'arrival_date_day_of_month', 
    'stays_in_weekend_nights', 'stays_in_weekdays_nights', 
    'days_in_waiting_list'
]

In [ ]:
desc_cats = df[categorical].astype(str).describe().T
desc_cats['frac_top'] = desc_cats['freq'] * 100 / df.shape[0]
desc_cats

count unique            top    freq   frac_top
hotel                 119390      2     City Hotel   79330  66.446101
is_canceled           119390      2              0   75166  62.958372
meal                  119390      5      Breakfast   92310  77.318033
city                  119390    178  Kota Denpasar   48590  40.698551
market_segment        119390      8      Online TA   56477  47.304632
distribution_channel  119390      5          TA/TO   97870   81.97504
is_repeated_guest     119390      2              0  115580  96.808778
reservation_status    119390      3      Check-Out   75166  62.958372
deposit_type          119390      3     No Deposit  104641  87.646369
agent                 119390    334            9.0   31961  26.770249
company               119390    353            nan  112593  94.306893
customer_type         119390      4       Personal   89613   75.05905

In [ ]:
desc_num = df[numerical].describe().round(3).T

for i, col in enumerate(numerical):
    desc_num.loc[col, 'no_unique'] = df[col].nunique()

desc_num['no_unique'] = desc_num['no_unique'].astype(int)
desc_num

count     mean     std   min    25%  \
adults                          119390.0    1.856   0.579  0.00   2.00   
children                        119386.0    0.104   0.399  0.00   0.00   
babies                          119390.0    0.008   0.097  0.00   0.00   
previous_cancellations          119390.0    0.087   0.844  0.00   0.00   
previous_bookings_not_canceled  119390.0    0.137   1.497  0.00   0.00   
booking_changes                 119390.0    0.221   0.652  0.00   0.00   
adr                             119390.0  101.831  50.536 -6.38  69.29   
required_car_parking_spaces     119390.0    0.063   0.245  0.00   0.00   
total_of_special_requests       119390.0    0.571   0.793  0.00   0.00   

                                   50%    75%     max  no_unique  
adults                           2.000    2.0    55.0         14  
children                         0.000    0.0    10.0          5  
babies                           0.000    0.0    10.0          5  
previous_cancellations           0.000    0.0    26.0         15  
previous_bookings_not_canceled   0.000    0.0    72.0         73  
booking_changes                  0.000    0.0    21.0         21  
adr                             94.575  126.0  5400.0       8879  
required_car_parking_spaces      0.000    0.0     8.0          5  
total_of_special_requests        0.000    1.0     5.0          6

In [ ]:
desc_time = df[time_related].describe().T

for i, col in enumerate(time_related):
    desc_time.loc[col, 'no_unique'] = df[col].nunique()

desc_time

count         mean         std     min     25%  \
lead_time                  119390.0   104.011416  106.863097     0.0    18.0   
arrival_date_year          119390.0  2018.156554    0.707476  2017.0  2018.0   
arrival_date_week_number   119390.0    27.165173   13.605138     1.0    16.0   
arrival_date_day_of_month  119390.0    15.798241    8.780829     1.0     8.0   
stays_in_weekend_nights    119390.0     0.927599    0.998613     0.0     0.0   
stays_in_weekdays_nights   119390.0     2.500302    1.908286     0.0     1.0   
days_in_waiting_list       119390.0     2.321149   17.594721     0.0     0.0   
arrival_date_month              NaN          NaN         NaN     NaN     NaN   

                              50%     75%     max  no_unique  
lead_time                    69.0   160.0   737.0      479.0  
arrival_date_year          2018.0  2019.0  2019.0        3.0  
arrival_date_week_number     28.0    38.0    53.0       53.0  
arrival_date_day_of_month    16.0    23.0    31.0       31.0  
stays_in_weekend_nights       1.0     2.0    19.0       17.0  
stays_in_weekdays_nights      2.0     3.0    50.0       35.0  
days_in_waiting_list          0.0     0.0   391.0      128.0  
arrival_date_month            NaN     NaN     NaN       12.0

## Observations

- Children, agent, company and adr bertipe data float, sehingga perlu disesuaikan tipe datanya.
- Terdapat tiga variable memiliki missing values yang cukup banyak, seperti city, agent and company.
- Ada beberapa variabel yang memiliki value yang tidak sesuai. Banyak data yang bernilai nol seperti adult, children dan babies, sehingga perlu diinvestigasi.


In [ ]:
df['adults'].value_counts()

2     89680
1     23027
3      6202
0       403
4        62
26        5
27        2
20        2
5         2
40        1
50        1
55        1
6         1
10        1
Name: adults, dtype: int64

In [ ]:
df['children'].value_counts()

0.0     110796
1.0       4861
2.0       3652
3.0         76
10.0         1
Name: children, dtype: int64

In [ ]:
df['babies'].value_counts()

0     118473
1        900
2         15
10         1
9          1
Name: babies, dtype: int64

Pada umumnya, jumlah pengunjung hotel sebanyak 4 orang. Dan jumlah adult yang bernilai nol dan jumlah anak yang berjumlah 10 dirasa tidak masuk akal.

In [ ]:
children_filter = (
    (df['children'] > 3) |
    (df['babies'] > 2)
)
df[children_filter]

hotel  is_canceled  lead_time  arrival_date_year  \
328    Resort Hotel            1         55               2017   
46619    City Hotel            0         37               2018   
78656    City Hotel            0         11               2017   

      arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
328            September                        29                         12   
46619              March                         3                         12   
78656           December                        42                         11   

       stays_in_weekend_nights  stays_in_weekdays_nights  adults  ...  \
328                          4                        10       2  ...   
46619                        0                         2       2  ...   
78656                        2                         1       1  ...   

       booking_changes  deposit_type agent company days_in_waiting_list  \
328                  2    No Deposit   8.0     NaN                    0   
46619                1    No Deposit   9.0     NaN                    0   
78656                1    No Deposit  95.0     NaN                    0   

      customer_type     adr  required_car_parking_spaces  \
328        Contract  133.16                            0   
46619      Personal   84.45                            0   
78656        Family   95.00                            0   

       total_of_special_requests  reservation_status  
328                            1             No-Show  
46619                          1           Check-Out  
78656                          0           Check-Out  

[3 rows x 29 columns]

Dikarenakan pengunjung hotel yang memiliki 10 anak terdapat 2 adult sehingga bisa jadi pengunjung yang tidak biasa pada umumnya. Dan dapat ditoleransikan.

In [ ]:
df['meal'].value_counts()

Breakfast     92310
Dinner        14463
No Meal       10650
Undefined      1169
Full Board      798
Name: meal, dtype: int64

In [ ]:
df['meal'].replace({'Undefined': 'No Meal'}, regex=True, inplace=True)
(df['meal'] == 'Undefined' ).sum()

0

Variabel meal yang memiliki value Undefined dapat digantikan dengan NoMeal

In [ ]:
zero_guest = (
    (df['adults'] == 0) & 
    (df['children'] == 0) & 
    (df['babies'] == 0) &
    (df['reservation_status'] != 'Canceled')
)
df[zero_guest].shape

(156, 29)

Ada 156 reservasi dengan tidak ada pengunjung.

In [ ]:
zero_adult = (
    (df['adults'] == 0) & 
    (df['reservation_status'] != 'Canceled') 
)

df[zero_adult].shape

(298, 29)

Ada 298 kamar yang tidak ada pengunjung adult, sehingga perlu diinvestigasi ulang.

In [ ]:
zero_night = (
    (df['stays_in_weekend_nights'] == 0) & 
    (df['stays_in_weekdays_nights'] == 0) &
    (df['reservation_status'] != 'Canceled')
)
df[zero_night].shape

(693, 29)

Terdapat 693 reservasi yang tidak menginap sama sekali.

## Handling Missing Values

In [ ]:
df['children'] = df['children'].fillna(0)
    
df['city'] = df['city'].fillna('Unknown')
    
df['agent'] = df['agent'].fillna(0)
    
df['company'] = df['company'].fillna(0)

df['adr'].fillna(0)

df['required_car_parking_spaces'].fillna(0)

df['total_of_special_requests'].fillna(0)

df['reservation_status'].dropna()

df.isnull().sum().sum()

0

Penjelasan
- Untuk variabel children dapat diisikan dengan nilai 0, karena banyak pengunjung yang tidak membawa anak.
- Dikarenakan terdapat 488 missing pada variabel city cukup sulit mendefinisikan nilainya, sehingga dapat diisikan dengan Unknown.
- Variabel agent dan company memiliki banyak missing value. Karena agent dan company dapat diisikan dengan nilai 0, atau dapat disimpulkan bahwa pengunjung tidak menggunakan agent atau company dalam melakukan booking.

##  Handle Duplicated Data

In [ ]:
df.duplicated().sum()

33261

Terdapat 33261 data yang memiliki data duplikat.

In [ ]:
df[df.duplicated(keep=False)].head(3)

hotel  is_canceled  lead_time  arrival_date_year  \
4   Resort Hotel            0         14               2017   
5   Resort Hotel            0         14               2017   
21  Resort Hotel            0         72               2017   

   arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
4           September                        27                          1   
5           September                        27                          1   
21          September                        27                          1   

    stays_in_weekend_nights  stays_in_weekdays_nights  adults  ...  \
4                         0                         2       2  ...   
5                         0                         2       2  ...   
21                        2                         4       2  ...   

    booking_changes  deposit_type  agent company days_in_waiting_list  \
4                 0    No Deposit  240.0     0.0                    0   
5                 0    No Deposit  240.0     0.0                    0   
21                1    No Deposit  250.0     0.0                    0   

   customer_type    adr  required_car_parking_spaces  \
4       Personal  98.00                            0   
5       Personal  98.00                            0   
21      Personal  84.67                            0   

    total_of_special_requests  reservation_status  
4                           1           Check-Out  
5                           1           Check-Out  
21                          1           Check-Out  

[3 rows x 29 columns]

Data duplikat dapat di drop, karena banyak sekali variabel yang bernilai sama.

In [ ]:
df = df.drop_duplicates()

print(f"Total data: {df.shape}")

Total data: (86129, 29)


In [ ]:
for col in ['children', 'agent', 'company']:
    df[col] = df[col].astype(int)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


Data yang tidak sesuai dengan tipe datanya telah dirubah, dari float menjadi integer.

In [ ]:
zero_guest = (
    (df['adults'] == 0) & 
    (df['children'] == 0) & 
    (df['babies'] == 0) &
    (df['reservation_status'] != 'Canceled')
)

zero_adult = (
    (df['adults'] == 0) & 
    (df['reservation_status'] != 'Canceled') 
)

zero_night = (
    (df['stays_in_weekend_nights'] == 0) & 
    (df['stays_in_weekdays_nights'] == 0) &
    (df['reservation_status'] != 'Canceled')
)

df1 = df.loc[(~zero_guest) | (~zero_adult)].copy()
df_cleaned = df1.loc[~zero_night].reset_index(drop=True).copy()
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85403 entries, 0 to 85402
Data columns (total 29 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   hotel                           85403 non-null  object 
 1   is_canceled                     85403 non-null  int64  
 2   lead_time                       85403 non-null  int64  
 3   arrival_date_year               85403 non-null  int64  
 4   arrival_date_month              85403 non-null  object 
 5   arrival_date_week_number        85403 non-null  int64  
 6   arrival_date_day_of_month       85403 non-null  int64  
 7   stays_in_weekend_nights         85403 non-null  int64  
 8   stays_in_weekdays_nights        85403 non-null  int64  
 9   adults                          85403 non-null  int64  
 10  children                        85403 non-null  int64  
 11  babies                          85403 non-null  int64  
 12  meal                            

Data seperti zero_guest, zero_adult dan zero_night kemudian dikeluarkan sehingga data dapat diproses lebih lanjut.

In [ ]:
df_cleaned.to_csv("cleaned_hotel_bookings_data.csv", index=False)